ethexe/ethereum/src/fallback_ws.rs

```diff
@@ -0,0 +1,156 @@
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use alloy::{
+    providers::WsConnect,
+    pubsub::{ConnectionHandle, PubSubConnect},
+    rpc::client::{ClientBuilder, RpcClient},
+    transports::{RpcError, TransportError, TransportErrorKind, TransportResult},
+};
+use anyhow::anyhow;
+use std::collections::VecDeque;
+use tokio::sync::Mutex;
+
+/// Builds the [`RpcClient`] for using [`alloy::providers::Provider`]
+/// with inner [`FallbackWs`] pubsub connect implementation.
+///
+/// ## Usage example:
+/// ```rust ignore
+/// let main_ws = "wss://";
+/// // Fallback  ws
+/// let public_ws = "wss://infura.io.public.rpc/...";
+///
+/// let client = rpc_client_with_fallback(String::new(), vec![String::new()])
+///     .await
+///     .unwrap();
+/// let provider = ProviderBuilder::default().connect_client(client);
+/// ```
+pub async fn rpc_client_with_fallback(
+    current: String,
+    fallbacks: Vec<String>,
+) -> Result<RpcClient, TransportError> {
+    let fallbacks: VecDeque<_> = fallbacks.into_iter().map(WsConnect::new).collect();
+
+    let fallback_ws = FallbackWs {
+        current: Mutex::new(WsConnect::new(current)),
+        fallbacks: Mutex::new(fallbacks),
+    };
+    ClientBuilder::default().pubsub(fallback_ws).await
+}
+
+/// [`FallbackWs`] is a wrapper around [`WsConnect`].
+/// It implements [`PubSubConnect`] trait to manage the ws connection in case when they dropped.
+pub struct FallbackWs {
+    current: Mutex<WsConnect>,
+    fallbacks: Mutex<VecDeque<WsConnect>>,
+}
+
+impl PubSubConnect for FallbackWs {
+    fn is_local(&self) -> bool {
+        false
+    }
+
+    async fn connect(&self) -> TransportResult<ConnectionHandle> {
+        self.current.lock().await.connect().await
+    }
+
+    async fn try_reconnect(&self) -> TransportResult<ConnectionHandle> {
+        let mut current = self.current.lock().await;
+        let mut fallbacks = self.fallbacks.lock().await;
+
+        // stores drop ws connections
+        let mut dropped_connections = Vec::new();
+        dropped_connections.push(current.clone());
+
+        loop {
+            let next_ws = fallbacks.pop_front().ok_or_else(|| {
+                RpcError::Transport(TransportErrorKind::Custom(
+                    anyhow!("all ws connections dropped, verify the network").into(),
+                ))
+            })?;
+
+            match next_ws.connect().await {
+                Ok(conn) => {
+                    tracing::trace!(
+                        previous_connection = %current.url(),
+                        new_connection = %next_ws.url(),
+                        "reconnecting to new web socket"
+                    );
+                    // Add dropped connections to fallbacks, assume that in future they may be valid
+                    fallbacks.extend(dropped_connections);
+                    *current = next_ws;
+                    return Ok(conn);
+                }
+                Err(err) => {
```

Comment:
You can save error to provide it later with `anyhow` in `RpcError::Transport` but it's very optional to implement